py/server/tests/test_table_listener.py

```diff
@@ -3,102 +3,228 @@
+                self.t_update.modified_prev.append(update.modified_prev())
 
-            def has_table_changed(self):
-                return self.table_update
+                # self.table_update.shifted = update.shifted
 
         listener = ListenerClass()
-        table_listener_handle = listen(self.table, listener)
-        time.sleep(2)
-        self.assertTrue(listener.has_table_changed())
-        table_listener_handle.deregister()
-
-    def test_shift_aware_listener_func(self):
-        def listener_func(update):
-            added_iterator = update.added.iterator()
-            while added_iterator.hasNext():
-                idx = added_iterator.nextLong()
-                ts = self.table.j_table.getColumnSource("Timestamp").get(idx)
-                x = self.table.j_table.getColumnSource("X").get(idx)
-                self.table_change_processed = ts or x
-
-            removed_iterator = update.removed.iterator()
-            while removed_iterator.hasNext():
-                idx = removed_iterator.nextLong()
-                ts = self.table.j_table.getColumnSource("Timestamp").getPrev(idx)
-                x = self.table.j_table.getColumnSource("X").getPrev(idx)
-                self.table_change_processed = ts or x
-
-        table_listener_handle = listen(self.table, listener_func)
+
+        table_listener_handle = listen(self.test_table, listener)
+        time.sleep(1)
+        pprint(listener.t_update.added)
+        pprint("----" * 10)
+        pprint(listener.t_update.removed)
+        self.check_result(listener.t_update.added, cols="X")
+        self.check_result(listener.t_update.removed)
+        self.check_result(listener.t_update.modified)
+        self.check_result(listener.t_update.modified_prev)
+
+        table_listener_handle.stop()
+
+    def test_listener_func(self):
+        replay_recorder = []
+        t_update = SimpleNamespace(added=None, removed=None, modified=None, modified_prev=None,
+                                       Shifted=None, modified_columns=None)
+
+        def listener_func(update, is_replay):
+            t_update.added = []
+            t_update.removed = []
+            t_update.modified = []
+            t_update.modified_prev = []
+            t_update.added.append(update.added())
+            t_update.removed.append(update.removed())
+            t_update.modified.append(update.modified())
+            t_update.modified_prev.append(update.modified_prev())
+            replay_recorder.append(is_replay)
+
+        table_listener_handle = TableListenerHandle(self.test_table, listener_func)
+        table_listener_handle.start(do_replay=True)
+        time.sleep(1)
+        self.assertGreaterEqual(len(t_update.added), 1)
+        self.check_result(t_update.added)
+        self.check_result(t_update.removed)
+        self.check_result(t_update.modified)
+        self.check_result(t_update.modified_prev)
+        self.assertFalse(all(replay_recorder))
```

Comment:
Not sure about the DRY comment. We want to check all the changes captured.

Reply:
I agree that you want to check everything.  I put the DRY comment because the code could be structured in a way that the same code doesn't appear multiple times.  E.g. a class to keep track of values and assert everything, with a common API to update values.  As the code morphs, we currently have to hope that the next guy remembers to make changes in all of the relevant places.  As the code currently is, some values are still not being checked in all unit tests (e.g. modified columns).  If we at least tested all of the values for all of the cases, it would probably be ok.